File: simplify-android/src/test/java/com/simplify/android/sdk/SimplifyTest.java
package com.simplify.android.sdk;


import android.app.Activity;
import android.content.Intent;
import android.os.Handler;
import android.os.Message;

import com.google.android.gms.wallet.FullWallet;
import com.google.android.gms.wallet.MaskedWallet;
import com.google.android.gms.wallet.WalletConstants;

import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class SimplifyTest {

    Simplify simplify;

    Simplify.AndroidPayCallback spyAndroidPayCallback;
    Intent mockResultData;

    @Before
    public void setUp() throws Exception {
        simplify = spy(new Simplify());

        spyAndroidPayCallback = spy(new Simplify.AndroidPayCallback() {
            @Override
            public void onReceivedMaskedWallet(MaskedWallet maskedWallet) {
            }

            @Override
            public void onReceivedFullWallet(FullWallet fullWallet) {
            }

            @Override
            public void onAndroidPayCancelled() {
            }

            @Override
            public void onAndroidPayError(int errorCode) {
            }
        });

        mockResultData = mock(Intent.class);
    }

    @Test
    public void testValidateApiKeyReturnsFalseWhenIncorrectFormat() {
        String badUuidEncoded = "bm8gYSB1dWlk";
        String badUuidDecoded = "not a uuid";
        String badKey = "lvpb_" + badUuidEncoded;

        doReturn(badUuidDecoded).when(simplify).base64Decode(badUuidEncoded);

        boolean returnValue = simplify.validateApiKey(badKey);
        assertFalse(returnValue);

        // missing prefix
        returnValue = simplify.validateApiKey(badUuidEncoded);
        assertFalse(returnValue);
    }

    @Test
    public void testValidateApiKeyReturnsTrueWithGoodKey() {
        String uuidEncoded = "M2E2YTJkOTctMDEwZS00MjViLWJhZWItZmI1Yjg1NTMxMDk3";
        String uuidDecoded = "3a6a2d97-010e-425b-baeb-fb5b85531097";
        String goodKey = "lvpb_" + uuidEncoded;

        doReturn(uuidDecoded).when(simplify).base64Decode(uuidEncoded);

        boolean returnValue = simplify.validateApiKey(goodKey);
        assertTrue(returnValue);
    }

    @Test
    public void testInitWithValidApiKeyIsSuccessful() {
        String apiKey = "lvpb_M2E2YTJkOTctMDEwZS00MjViLWJhZWItZmI1Yjg1NTMxMDk3";

        doReturn(true).when(simplify).validateApiKey(any());

        simplify.setApiKey(apiKey);

        assertEquals(simplify.apiKey, apiKey);
    }

    @Test
    public void testInitWithInvalidApiKeyThrowsError() {
        String key = "invalid";

        doReturn(false).when(simplify).validateApiKey(key);

        try {
            simplify.setApiKey(key);
            fail();
        } catch (Exception e) {
            assertTrue(e instanceof IllegalArgumentException);
        }
    }

    @Test
    public void testGetUrlWithLiveApiKey() {
        simplify.apiKey = "lvpb_123456789";

        assertEquals(simplify.getUrl(), Constants.API_BASE_LIVE_URL);
    }

    @Test
    public void testGetUrlWithSandboxApiKey() {
        simplify.apiKey = "sbpb_123456789";

        assertEquals(simplify.getUrl(), Constants.API_BASE_SANDBOX_URL);
    }

    @Test
    public void testMissingAndroidPayCallbackThrowsException() {
        try {
            simplify.handleAndroidPayResult(Simplify.REQUEST_CODE_MASKED_WALLET, Activity.RESULT_OK, mockResultData, null);
            fail();
        } catch (Exception e) {
            assertTrue(e instanceof IllegalArgumentException);
        }
    }

    @Test
    public void testUnrecognizedRequestCodeReturnsFalse() {
        int unrecognizedRequestCode = 123;

        boolean handled = simplify.handleAndroidPayResult(unrecognizedRequestCode, Activity.RESULT_OK, mockResultData, spyAndroidPayCallback);

        assertFalse(handled);
    }

    @Test
    public void testResultCancelledCallsCancelledMethodOnCallback() {
        simplify.handleAndroidPayResult(Simplify.REQUEST_CODE_MASKED_WALLET, Activity.RESULT_CANCELED, mockResultData, spyAndroidPayCallback);

        verify(spyAndroidPayCallback).onAndroidPayCancelled();
    }

    @Test
    public void testErrorResultCallsErrorMethodOnCallback() {
        int errorResultCode = 100;

        // mock getting error code out of intent data
        when(mockResultData.getIntExtra(anyString(), anyInt())).thenReturn(errorResultCode);

        simplify.handleAndroidPayResult(Simplify.REQUEST_CODE_MASKED_WALLET, errorResultCode, mockResultData, spyAndroidPayCallback);

        verify(spyAndroidPayCallback).onAndroidPayError(errorResultCode);
    }

    @Test
    public void testMaskedWalletResultReturnsWalletToCallback() {
        MaskedWallet maskedWallet = null;

        when(mockResultData.hasExtra(WalletConstants.EXTRA_MASKED_WALLET)).thenReturn(true);
        when(mockResultData.getParcelableExtra(WalletConstants.EXTRA_MASKED_WALLET)).thenReturn(maskedWallet);

        simplify.handleAndroidPayResult(Simplify.REQUEST_CODE_MASKED_WALLET, Activity.RESULT_OK, mockResultData, spyAndroidPayCallback);

        verify(spyAndroidPayCallback).onReceivedMaskedWallet(maskedWallet);
    }

    @Test
    public void testFullWalletResultReturnsWalletToCallback() {
        FullWallet fullWallet = null;

        when(mockResultData.hasExtra(WalletConstants.EXTRA_FULL_WALLET)).thenReturn(true);
        when(mockResultData.getParcelableExtra(WalletConstants.EXTRA_FULL_WALLET)).thenReturn(fullWallet);

        simplify.handleAndroidPayResult(Simplify.REQUEST_CODE_FULL_WALLET, Activity.RESULT_OK, mockResultData, spyAndroidPayCallback);

        verify(spyAndroidPayCallback).onReceivedFullWallet(fullWallet);
    }

    @Test
    public void testHandlerMessageContainingThrowableCallsOnErrorCallback() {
        Exception exception = new Exception();
        CardToken.Callback mockCallback = mock(CardToken.Callback.class);

        boolean returnValue = simplify.handleCreateCardTokenCallbackMessage(mockCallback, exception);

        verify(mockCallback).onError(exception);
        assertTrue(returnValue);
    }

    @Test
    public void testHandlerMessageContainingTCardTokenCallsOnSuccessCallback() {
        CardToken mockCardToken = mock(CardToken.class);
        CardToken.Callback mockCallback = mock(CardToken.Callback.class);

        boolean returnValue = simplify.handleCreateCardTokenCallbackMessage(mockCallback, mockCardToken);

        verify(mockCallback).onSuccess(mockCardToken);
        assertTrue(returnValue);
    }

    @Test
    public void testCreateCardTokenRunnableSendsThrowableMessageOnError() throws Exception {
        Handler mockHandler = mock(Handler.class);
        Message message = new Message();
        Exception exception = new Exception();

        when(mockHandler.obtainMessage()).thenReturn(message);
        doThrow(exception).when(simplify).executeCreateCardToken(any(), any());

        simplify.runCreateCardToken(new Card(), new Secure3DRequestData(), mockHandler);

        assertEquals(message.obj, exception);
        verify(mockHandler).sendMessage(message);
    }

    @Test
    public void testCreateCardTokenRunnableSendsCardTokenMessageOnSuccess() throws Exception {
        Handler mockHandler = mock(Handler.class);
        Message message = new Message();
        CardToken cardToken = mock(CardToken.class);

        when(mockHandler.obtainMessage()).thenReturn(message);
        doReturn(cardToken).when(simplify).executeCreateCardToken(any(), any());

        simplify.runCreateCardToken(new Card(), new Secure3DRequestData(), mockHandler);

        assertEquals(message.obj, cardToken);
        verify(mockHandler).sendMessage(message);
    }
}

File: simplify-android/src/test/kotlin/com/simplify/android/sdk/SimplifyTest.kt
package com.simplify.android.sdk

import android.app.Activity
import android.content.Intent
import com.nhaarman.mockitokotlin2.*
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.MockitoAnnotations
import org.mockito.Spy
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import java.lang.Exception
import java.lang.IllegalArgumentException
import java.lang.RuntimeException
import kotlin.test.*

@RunWith(RobolectricTestRunner::class)
@Config(manifest = Config.NONE)
class SimplifyTest {

    private lateinit var simplify: Simplify

    @Mock
    private lateinit var mockComms: SimplifyComms

    @Spy
    private lateinit var spy3DSCallback: SimplifySecure3DCallback


    @Before
    fun setUp() {
        MockitoAnnotations.initMocks(this)

        simplify = Simplify("sbpb_M2E2YTJkOTctMDEwZS00MjViLWJhZWItZmI1Yjg1NTMxMDk3")
        simplify.comms = mockComms
    }

    @After
    fun tearDown() {
        reset(mockComms,  spy3DSCallback)
    }

    @Test
    fun testApiKeyPassedInConstructorIsValidated() {
        val goodPrefix = "lvpb_"
        val badKeyId = "bm90IGEgdXVpZA==" // "not a uuid"

        try {
            Simplify(goodPrefix + badKeyId)
            fail("Should have thrown exception with invalid key format in constructor")
        } catch (e: Exception) {
            assertTrue(e is IllegalArgumentException)
        }
    }

    @Test
    fun testSetApiKeyThrowsExceptionWhenIncorrectFormat() {
        val goodPrefix = "lvpb_"
        val badKeyId = "bm90IGEgdXVpZA==" // "not a uuid"

        try {
            simplify.apiKey = goodPrefix + badKeyId
            fail("Should have thrown exception with invalid key format")
        } catch (e: Exception) {
            assertTrue(e is IllegalArgumentException)
        }
    }

    @Test
    fun testSetApiKeyThrowsExceptionOnBadPrefix() {
        val badPrefix = "bogus_"
        val goodKeyId = "M2E2YTJkOTctMDEwZS00MjViLWJhZWItZmI1Yjg1NTMxMDk3"

        try {
            simplify.apiKey = badPrefix + goodKeyId
            fail("Should have thrown exception with invalid key prefix")
        } catch (e: Exception) {
            assertTrue(e is IllegalArgumentException)
        }
    }

    @Test
    fun testSetApiKeyWorksWithGoodKey() {
        val goodPrefix = "lvpb_"
        val goodKeyId = "M2E2YTJkOTctMDEwZS00MjViLWJhZWItZmI1Yjg1NTMxMDk3"

        try {
            simplify.apiKey = goodPrefix + goodKeyId
            assertEquals(goodPrefix + goodKeyId, simplify.apiKey)
        } catch (e: Exception) {
            fail("Should have passed all validation rules")
        }
    }

    @Test
    fun testCorrectBaseUrlUsedForKeyType() {
        val livePrefix = "lvpb_"
        val sandboxPrefix = "sbpb_"
        val keyId = "M2E2YTJkOTctMDEwZS00MjViLWJhZWItZmI1Yjg1NTMxMDk3"

        simplify.apiKey = sandboxPrefix + keyId
        simplify.createCardToken(SimplifyMap(), null, mock())

        simplify.apiKey = livePrefix + keyId
        simplify.createCardToken(SimplifyMap(), null, mock())

        argumentCaptor<SimplifyRequest>().apply {
            verify(mockComms, times(2)).runSimplifyRequest(capture(), any())

            assertTrue(firstValue.url.startsWith(Simplify.API_BASE_SANDBOX_URL))
            assertTrue(secondValue.url.startsWith(Simplify.API_BASE_LIVE_URL))
        }
    }

    @Test
    fun testRequestBuiltCorrectlyForCreateCardToken() {
        val apiKey = "lvpb_M2E2YTJkOTctMDEwZS00MjViLWJhZWItZmI1Yjg1NTMxMDk3"

        val card = SimplifyMap()
                .set("number", "5555555555554444")
                .set("expMonth", "01")
                .set("expYear", "50")
                .set("cvc", "123")

        val secure3DRequestData = SimplifyMap()
                .set("paReq", "some data")

        simplify.apiKey = apiKey
        simplify.createCardToken(card, null, mock())
        simplify.createCardToken(card, secure3DRequestData, mock())

        argumentCaptor<SimplifyRequest>().apply {
            verify(mockComms, times(2)).runSimplifyRequest(capture(), any())

            // common assertions
            assertTrue(firstValue.url.endsWith(Simplify.API_PATH_CARDTOKEN))
            assertEquals(SimplifyRequest.Method.POST, firstValue.method)
            assertEquals(apiKey, firstValue.payload["key"])
            assertEquals(card["number"], firstValue.payload["card.number"])

            // first request, missing 3DS data
            assertNull(firstValue.payload["secure3DRequestData"])

            // second request, includes 3DS data
            assertEquals(secure3DRequestData["paReq"], secondValue.payload["secure3DRequestData.paReq"])
        }
    }

    @Test
    fun test3DSUnrecognizedRequestCodeReturnsFalse() {
        val unrecognizedRequestCode = 123

        val result = Simplify.handle3DSResult(unrecognizedRequestCode, Activity.RESULT_OK, mock(), mock())

        assertFalse("Should return false on unrecognized activity request code") { result }
    }

    @Test
    fun test3DSResultCancelledCallsCancelledMethodOnCallback() {
        val result = Simplify.handle3DSResult(Simplify.REQUEST_CODE_3DS, Activity.RESULT_CANCELED, mock(), spy3DSCallback)

        verify(spy3DSCallback).onSecure3DCancel()
        assertTrue("Should return true when handling cancel event") { result }
    }

    @Test
    fun testErrorReading3DSResultCallsErrorOnCallback() {
        val mockData: Intent = mock()
        whenever(mockData.getStringExtra(SimplifySecure3DActivity.EXTRA_RESULT)).doThrow(RuntimeException())

        val result = Simplify.handle3DSResult(Simplify.REQUEST_CODE_3DS, Activity.RESULT_OK, mockData, spy3DSCallback)

        verify(spy3DSCallback).onSecure3DError(any())
        assertTrue("Should return true when handling error event") { result }
    }

    @Test
    fun testMissingSecure3dResponseDataCallsErrorOnCallback() {
        val mockData: Intent = mock()
        whenever(mockData.getStringExtra(SimplifySecure3DActivity.EXTRA_RESULT)).doReturn("{\"missing\":\"data\"}")

        val result = Simplify.handle3DSResult(Simplify.REQUEST_CODE_3DS, Activity.RESULT_OK, mockData, spy3DSCallback)

        verify(spy3DSCallback).onSecure3DError(any())
        assertTrue("Should return true when handling error event") { result }
    }

    @Test
    fun test3DSErrorInResponseCallsErrorOnCallback() {
        val expectedErrorMessage = "3ds error message"

        val mockData: Intent = mock()
        whenever(mockData.getStringExtra(SimplifySecure3DActivity.EXTRA_RESULT)).doReturn("{\"secure3d\":{\"error\":{\"message\":\"$expectedErrorMessage\"}}}")

        val result = Simplify.handle3DSResult(Simplify.REQUEST_CODE_3DS, Activity.RESULT_OK, mockData, spy3DSCallback)

        verify(spy3DSCallback).onSecure3DError(expectedErrorMessage)
        assertTrue("Should return true when handling error event") { result }
    }

    @Test
    fun test3DSSuccessInResponseCallsCompleteOnCallback() {
        val authenticated = true

        val mockData: Intent = mock()
        whenever(mockData.getStringExtra(SimplifySecure3DActivity.EXTRA_RESULT)).doReturn("{\"secure3d\":{\"authenticated\":$authenticated}}")

        val result = Simplify.handle3DSResult(Simplify.REQUEST_CODE_3DS, Activity.RESULT_OK, mockData, spy3DSCallback)

        verify(spy3DSCallback).onSecure3DComplete(authenticated)
        assertTrue("Should return true when handling complete event") { result }
    }

    @Test
    fun testStart3DSActivityThrowsExceptionIfMissingData() {
        try {
            Simplify.start3DSActivity(mock(), mock())
            fail("Should have thrown an exception when no card.secure3DData provided")
        } catch (e: Exception) {
            assertTrue(e is IllegalArgumentException)
        }
    }

    @Test
    fun testStart3DSActivityUsesInternalRequestCode() {
        val mockActivity: Activity = mock()
        val cardToken = SimplifyMap()
                .set("card.secure3DData", SimplifyMap())

        Simplify.start3DSActivity(mockActivity, cardToken)

        verify(mockActivity).startActivityForResult(any(), eq(Simplify.REQUEST_CODE_3DS))
    }
}